Mathlib/Tactic/RelCongr/Core.lean

```diff
@@ -0,0 +1,479 @@
+              -- See if the term is `a = b` and the goal is `a ∼ b` or `b ∼ a`, with `∼` reflexive
+              let m ← mkFreshExprMVar none
+              g.exact (← mkAppOptM ``Eq.subst #[h, m])
+              g.rfl
+        return
+      catch _ => s.restore
+    throwError "rel_assumption failed"
+
+syntax "rel_congr_discharger" : tactic
+
+/-- The `rel_congr` tactic applies "relational congruence" rules, reducing a relational goal
+between a LHS and RHS matching the same pattern to relational subgoals between the differing
+inputs to the pattern.  For example,
+```
+example {a b x c d : ℝ} (h1 : a + 1 ≤ b + 1) (h2 : c + 2 ≤ d + 2) :
+    x ^ 2 * a + c ≤ x ^ 2 * b + d := by
+  rel_congr
+  · linarith
+  · linarith
+```
+This example has the goal of proving the relation `≤` between a LHS and RHS both of the pattern
+```
+x ^ 2 * ?_ + ?_
+```
+(with inputs `a`, `c` on the left and `b`, `d` on the right); after the use of
+`rel_congr`, we have the simpler goals `a ≤ b` and `c ≤ d`.
+
+A pattern can be provided explicitly; this is useful if a non-maximal match is desired:
+```
+example {a b c d x : ℝ} (h : a + c + 1 ≤ b + d + 1) :
+    x ^ 2 * (a + c) + 5 ≤ x ^ 2 * (b + d) + 5 := by
+  rel_congr x ^ 2 * ?_ + 5
+  linarith
+```
+
+The "relational congruence" rules used are the library lemmas which have been tagged with the
+attribute `@[rel_congr]`.  For example, the first example constructs the proof term
+```
+add_le_add (mul_le_mul_of_nonneg_left _ (pow_bit0_nonneg x 1)) _
+```
+using the relational congruence lemmas `add_le_add` and `mul_le_mul_of_nonneg_left`.
+
+The tactic attempts to discharge side goals to these "relational congruence" lemmas (such as the
+side goal `0 ≤ x ^ 2` in the above application of `mul_le_mul_of_nonneg_left`) using the tactic
+`rel_congr_discharger`, which wraps `positivity` but can also be extended. Side goals not discharged
+in this way are left for the user. -/
+elab "rel_congr" template:(colGt term)?
+    withArg:((" with " (colGt binderIdent)+)?) : tactic => do
+  let g ← getMainGoal
+  g.withContext do
+  let .app (.app _rel lhs) _rhs ← withReducible g.getType'
+    | throwError "rel_congr failed, not a relation"
+  -- Elaborate the template (e.g. `x * ?_ + _`), if the user gave one
+  let template ← template.mapM fun e => do
+    Term.elabTerm e (← inferType lhs)
+  -- Get the names from the `with x y z` list
+  let names := (withArg.raw[1].getArgs.map TSyntax.mk).toList
+  -- The core tactic `Lean.MVarId.relCongr` will be run with discharger `rel_congr_discharger`
+  let disch g := Term.TermElabM.run' do
+    let [] ← Tactic.run g <| evalTactic (Unhygienic.run `(tactic| rel_congr_discharger))
+      | failure
+  -- The core tactic `Lean.MVarId.relCongr` will be run with "assumption tactic"
+  -- consisting of running `Lean.MVarId.relAssumption` (trying a term together with limited
+  -- forward-reasoning on that term) on each nontrivial hypothesis.
+  let assum g := do
```

Comment:
I have started to factor out the used lemmas, so that one has just a list of maps `Expr -> MVarId -> MetaM Unit` and this can be passed into the `relAssumption` function. Feel free to revert if there is a smarter solution.

Reply:
This is now done, thanks to Mario.